klaus/contrib/autodiscover.py

```diff
@@ -0,0 +1,194 @@
+    automatically when any changes are made. if `recursive=True`, klaus will
+    watch every subdirectory in the given directories, and discover all git
+    repositories therein. This simplifies administration significantly, but may
+    cause problems with large repositories, as the underlying file watch
+    mechanisms don't expose a recursion depth parameter on any of the common
+    platforms (Linux, win32, OS X).
+    """
+    def __init__(self, basedirs, site_name, use_smarthttp, **kwargs):
+        recursive = kwargs.pop('discover_recursive', False)
+        super(KlausAutoDiscover, self).__init__([], site_name, use_smarthttp, **kwargs)
+        self._lock = threading.RLock()
+        self._basedirs = basedirs
+        self._observer = Observer()
+
+        # First find and add all the existing repos in the given base directory.
+        for directory in basedirs:
+            for fspath in os.listdir(directory):
+                path = os.path.join(directory, fspath)
+                self.add_repo(path)
+
+        # Then create a handler to notify us of any changes within those dirs.
+        fs_event_handler = KlausFSEventHandler(self)
+        for directory in basedirs:
+            self._observer.schedule(fs_event_handler, directory, recursive=recursive)
+
+        self._observer.start()
+
+    def add_repo(self, directory):
+        """
+        Attempts to register `directory` with the klaus app as a git repository.
+        If `directory` is not a git repo, this method does nothing.
+        """
+        with self._lock:
+            try:
+                repo = FancyRepo(directory)
+            except dulwich.errors.NotGitRepository:
+                logger.debug("Dulwich says '%s' is not a git repository", directory)
+                return
+
+            if self.repo_map.get(repo.name):
+                logger.debug("skipping repo with duplicate name:%s", repo)
+                return
+            logger.info("Adding repository, '%s' at '%s'", repo.name, directory)
+            self.repo_map[repo.name] = repo
+
+    def remove_repo(self, directory):
+        """
+        given a filesystem path, `directory` attempts to remove the
+        corresponding repository from the list of watched repositories.
+        """
+        with self._lock:
+            key_map = dict(zip(
+                (repo.path for repo in self.repo_map.values()),
+                self.repo_map)
+            )
+            try:
+                name = key_map[directory]
+                del self.repo_map[name]
+                return True
+            except KeyError:
+                logger.debug(
+                    "couldn't delete '%s'; not an existing repository",
+                    directory
+                )
+                pass
+
+    def __del__(self):
+        self._observer.stop()
+        self._observer.join()
+
+
+def _app(htdigest_file=None):
+    return make_app(
```

Comment:
Putting the code common to the both cases into a function is a good idea, but the pattern used elsewhere (`wsgi.py` and `wsgi_autoreload.py`) is to repeat it. Can we do this particular cleanup after in a different PR?